chain/chunks/src/lib.rs

```diff
@@ -603,32 +603,25 @@ impl ShardsManager {
                 continue;
             }
 
-            let need_to_fetch_part = if request_full || seal.contains_part_ord(&part_ord) {
-                true
+            // Note: If request_from_archival is true, we potentially call
+            // get_part_owner unnecessarily.  It’s probably not worth optimising
+            // though unless you can think of a concise way to do it.
+            let part_owner = self.runtime_adapter.get_part_owner(&epoch_id, part_ord)?;
```

Comment:
Given that `get_part_owner()` can fail and return an error, could this be a change in the function's behaviour.

Reply:
I don’t think so.  The only place this can fail is in this method:

```rust
pub fn get_epoch_info(&self, epoch_id: &EpochId) -> Result<Arc<EpochInfo>, EpochError> {
    self.epochs_info.get_or_try_put(epoch_id.clone(), |epoch_id| {
        self.store
            .get_ser(DBCol::EpochInfo, epoch_id.as_ref())?
            .ok_or_else(|| EpochError::EpochOutOfBounds(epoch_id.clone()))
    })
}
```

Either there’s database read failure (in which case the issue exists
regardless and eventually something else will fail) or we asked about
invalid epoch id.  The latter cannot happen since we got the id via:

```rust
let epoch_id = self.runtime_adapter.get_epoch_id_from_prev_block(ancestor_hash)?;
```

and EpochInfo isn’t garbage collected (and even if it was something we
care about wouldn’t be collected).

Reply:
I guess somewhat of a premature optimization, but might it be the case that checking for `request_full` and `seal.contains_part_ord` first makes the overall code run much faster (e.g. `request_full` might be `true` a lot of the time?)

Either way it seems like both this and akhi’s concern could be resolved by writing these checks as such, so that short-circuiting can take care of running `get_part_owner` at the same time as it was before. Giving the name to the condition also helps with this code being self-documenting.

```rust
let need_to_fetch = request_full || seal.contains_part_ord(&part_ord) || me.map(|me| {
    self.runtime_adapter.get_part_owner(&epoch_id, part_ord)? == me
}).unwrap_or(false)); // `Option::is_some_and` would work great here, but isn’t yet stable

// or

let need_to_fetch = request_full 
    || seal.contains_part_ord(&part_ord) 
    || (me.is_some()
    && Some(&self.runtime_adapter.get_part_owner(&epoch_id, part_ord)?) == me);

// -------

if !need_to_fetch { 
    continue;
}
```

Reply:
In this instance, if `request_from_archival` was false, we would have to `get_part_owner` twice.  The second time happening for the condition on line 617.